src/zkevm_specs/evm/execution/shr.py

```diff
@@ -0,0 +1,168 @@
+from ...util import FQ, N_BYTES_U64, RLC
+from ..instruction import Instruction, Transition
+from ..typing import Sequence
+
+
+def shr(instruction: Instruction):
+    opcode = instruction.opcode_lookup(True)
+
+    a = instruction.stack_pop()
+    shift = instruction.stack_pop()
+    b = instruction.stack_push()
+
+    (
+        a64s,
+        b64s,
+        a64s_lo,
+        a64s_hi,
+        shf_div64,
+        shf_mod64,
+        shf_lt256,
+        p_lo,
+        p_hi,
+    ) = gen_witness(instruction, a, shift)
+    check_witness(
+        instruction,
+        a,
+        shift,
+        b,
+        a64s,
+        b64s,
+        a64s_lo,
+        a64s_hi,
+        shf_div64,
+        shf_mod64,
+        shf_lt256,
+        p_lo,
+        p_hi,
+    )
+
+    instruction.step_state_transition_in_same_context(
+        opcode,
+        rw_counter=Transition.delta(2),
+        program_counter=Transition.delta(1),
+        stack_pointer=Transition.delta(1),
+    )
+
+
+def check_witness(
+    instruction: Instruction,
+    a: RLC,
+    shift: RLC,
+    b: RLC,
+    a64s: Sequence[FQ],
+    b64s: Sequence[FQ],
+    a64s_lo: Sequence[FQ],
+    a64s_hi: Sequence[FQ],
+    shf_div64,
+    shf_mod64,
+    shf_lt256,
+    p_lo,
+    p_hi,
+):
+    # Set `shf_real_div64` to an overflow value of 4 for merge contraints.
+    shf_real_div64 = instruction.select(shf_lt256, shf_div64, FQ(4))
```

Comment:
we don't need `shf_real_div64`. The idea is that we can disregard whether shift is overflowed, and just take the lower 8 bit value to prove the shift right. And we only use the value `b` to push to stack when `shift[1:31] == 0`.